### PR TITLE
Adding statistics collection support

### LINKED ISSUE #805: Support for Statistical Checkers
Statistical checkers require 2-phase analysis. In the 1st, statistics collection-phase, the source code is analyzed and information is collected about the API usages. These  statistics are stored into collection directory. In the 2nd, analysis phase, the collected information is used to find bugs in the code (API usages that are non-conform to the applied coding style).

CodeChecker will support this 2-phase analysis.
Checkers that benefit from the support are api.SpecialReturnValue, api.UncheckedReturnValue.

**Add  `--statistics-collect` option to `analyze` which switches on [experimental statistics collection](https://github.com/baloghadamsoftware/clang/tree/ReturnCheckers) checkers:**

 * `statisticsCollector.ReturnValueCheck`
 * `statisticsCollector.SpecialReturnValue`

The option will only be shown if the checkers are available.

The statistics will be stored into in `<REPORT-DIR>/statistics` directory.

If any of the 
* api.SpecialReturnValue
* api.UncheckedReturnValue

checkers are switched on, look for statistics file in the `<REPORT-DIR>/statistics` directory.

**Introduce a new optional analyze argument `--preanalize-dir <PREANALYZE-DIR>` which is a <REPORT_DIR> with the CTU and statistics pre-analysis results.** 

Using this parameter it will be possible to share the pre-analysis results between multiple analysis runs. So if this parameter is given, the `--ctu-analyze` and the `ctu-on-the-fly`, and the statistical checkers will search for the CTU and statistics interim files in `<PRE-ANALYZE-DIR>/ctu` and `<PRE-ANALYZE-DIR>/statistics` directories.

### LINKED ISSUE #1420: Build action map is created twice?
When both clangsa and tidy is turned on I receive the following debug warnings:
```
Multiple entires in compile database with the same (source, target) pair:
```
But the compile command json does not contain such multiple entries.
Perhaps we call the function `def create_actions_map(actions, manager)` twice on the same object.

### PR DESCRIPTION
**Pull request by gyorb**
* introduce new pre analysis step
  - already existing CTU
  - new statistic collector checkers (only collector checkers are enabled)

* new arguments (only active if the specific checkers are available)
 - --stats (collect statistics and analyze, enable statistics checkers)
 - --stats-collect (only collect statistics)
 - --stats-use (only analyze using the previously generated statistics)

* TODO yaml output generation should be done with a python yaml module

Resolves #805
Resolves #1420
Supersedes #967.

### CODE REVIEW

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py [hunk 1]
```diff
@@ -20,42 +20,46 @@
...
 
+def parse_checkers(clangsa_output):
+    """
+    Parse clang static analyzer checkers list output.
+    Return a list of (checker name, description) tuples.
+    """
+
+    # Checker name and description in one line.
+    pattern = re.compile(
+        r'^\s\s(?P<checker_name>\S*)\s*(?P<description>.*)')
```
Comment (csordasmarton) [inline libcodechecker/analyze/analyzers/analyzer_clangsa.py]: What if it contains more than two spaces at the beginning? Wouldn't it be better to write `^\s*`?

File: libcodechecker/libhandlers/analyze.py [hunk 2]
```diff
@@ -297,6 +296,46 @@ def add_arguments_to_parser(parser):
...
+            "These arguments are only available if the Clang Static Analyzer "
+            "supports Statistics-based analysis "
+            "(e.g. statisticsCollector.ReturnValueCheck,"
+            "statisticsCollector.SpecialReturnValue checkers are available).")
+        stat_opts.add_argument('--stats-collect', '--stats-collect',
+                               action='store',
+                               default=None,
+                               dest='stats_collect',
+                               help="Perform the first, 'collect' phase of "
+                                    "Statistical analysis. This phase"
```
Comment (csordasmarton) [inline libcodechecker/libhandlers/analyze.py]: Please write a space before the last word of every such lines.
```bash
"Statistical analysis. This phase "
```
Plese check the other help messages and places too.

File: libcodechecker/libhandlers/analyze.py [hunk 3]
```diff
@@ -297,6 +296,46 @@ def add_arguments_to_parser(parser):
                                    "same translation unit without "
                                    "Cross-TU enabled.")
 
+    if host_check.is_statistics_capable():
+        stat_opts = parser.add_argument_group(
+            "statistics analysis arguments",
+            "These arguments are only available if the Clang Static Analyzer "
+            "supports Statistics-based analysis "
+            "(e.g. statisticsCollector.ReturnValueCheck,"
```
Comment (csordasmarton) [inline libcodechecker/libhandlers/analyze.py]: Space after `,`

File: libcodechecker/host_check.py [hunk 4]
```diff
@@ -25,6 +27,37 @@ def is_ctu_capable():
...
+
+    check_env = analyzer_env.get_check_env(context.path_env_extra,
+                                           context.ld_lib_path_extra)
+
+    checkers = analyzer.get_analyzer_checkers(clangsa_cfg, check_env)
+
+    import re
+    stat_checkers_pattern = re.compile(r'.+statisticscollector.+')
+
+    stats_checkers = []
```
Comment (csordasmarton) [inline libcodechecker/host_check.py]: I think you do not use this variable.

File: libcodechecker/host_check.py [hunk 5]
```diff
@@ -25,6 +27,37 @@ def is_ctu_capable():
...
+    analyzer = analyzer_types.construct_analyzer_type(analyzer,
+                                                      clangsa_cfg,
+                                                      None)
+
+    check_env = analyzer_env.get_check_env(context.path_env_extra,
+                                           context.ld_lib_path_extra)
+
+    checkers = analyzer.get_analyzer_checkers(clangsa_cfg, check_env)
+
+    import re
```
Comment (csordasmarton) [inline libcodechecker/host_check.py]: What is the main reason you imported this module here not top of the file?

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py (hunk 1 above)
Comment (gyorb) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: I think we can change it (the output did not change in the last years), but I would do it in a separate pull request.

File: docs/user_guide.md [hunk 6]
```diff
@@ -681,6 +691,38 @@ cross translation unit analysis arguments:
...
+  --stats-collect STATS_COLLECT, --stats-collect STATS_COLLECT
+                        Perform the first, 'collect' phase of Statistical
+                        analysis. This phase generates extra files needed by
+                        statistics analysis, and puts them into
+                        '<OUTPUT_DIR>'. NOTE: If this argument is present,
+                        CodeChecker will NOT execute the analyzers! (default:
+                        None)
+  --stats-use STATS_DIR, --stats-use STATS_DIR
+                        Use the previously generated statistics results for
+                        the analysis from the given '<STATISTICS_DIR>'.
```
[the remark this answers is not in the document]
Comment (gyorb) [reply docs/user_guide.md]: I would keep it in sync with the value printed by argparse `STATS_DIR`

File: libcodechecker/libhandlers/analyze.py [hunk 7]
```diff
@@ -297,6 +296,46 @@ def add_arguments_to_parser(parser):
...
+    if host_check.is_statistics_capable():
+        stat_opts = parser.add_argument_group(
+            "statistics analysis arguments",
+            "These arguments are only available if the Clang Static Analyzer "
+            "supports Statistics-based analysis "
+            "(e.g. statisticsCollector.ReturnValueCheck, "
+            "statisticsCollector.SpecialReturnValue checkers are available).")
+        stat_opts.add_argument('--stats-collect', '--stats-collect',
+                               action='store',
+                               default=None,
```
[the remark this answers is not in the document]
Comment (gyorb) [reply libcodechecker/libhandlers/analyze.py]: done

File: libcodechecker/libhandlers/analyze.py [hunk 8]
```diff
@@ -297,6 +296,46 @@ def add_arguments_to_parser(parser):
...
+                                    "statistics analysis, and "
+                                    "puts them into "
+                                    "'<OUTPUT_DIR>'."
+                                    " NOTE: If this argument is present, "
+                                    "CodeChecker will NOT execute the "
+                                    "analyzers!")
+
+        stat_opts.add_argument('--stats-use', '--stats-use',
+                               action='store',
+                               default=None,
```
[the remark this answers is not in the document]
Comment (gyorb) [reply libcodechecker/libhandlers/analyze.py]: done

**Pull request merged**